expect('json') more strict on structure than expectJSON() was

## v0
```
frisby.create('gets existing basket by id')
  .addHeaders(authHeader)
  .get(REST_URL + '/basket/1')
  .expectStatus(200)
  .expectHeaderContains('content-type', 'application/json')
  .expectJSON('data', {id: 1})
  .toss()
```
works just fine.

## v2
```
  it('gets existing basket by id', function (done) {
    frisby.get(REST_URL + '/basket/1', {headers: authHeader})
      .expect('status', 200)
      .expect('header', 'content-type', /application\/json/)
      .expect('json', 'data', {id: 1})
      .done(done)
  })
```
fails with
```
Message:
     Expected 'AssertionError: { id: 1 } deepEqual { id: 1,
  coupon: null,
  createdAt: '2017-07-26T12:52:01.000Z',
  updatedAt: '2017-07-26T12:52:01.000Z',
  UserId: 1,
  produc
    at jsonAssertion (C:\Data\Github\juice-shop\node_modules\frisby\src\frisby\expects.js:68:14)
    at Object.withPath (C:\Data\Github\juice-shop\node_modules\frisby\src\frisby\utils.js:68:12)
    at FrisbySpec.json (C:\Data\Github\juice-shop\node_modules\frisby\src\frisby\expects.js:67:11)
    at FrisbySpec._addExpect (C:\Data\Github\juice-shop\node_modules\frisby\src\frisby\spec.js:385:23)
    at then (C:\Data\Github\juice-shop\node_modules\frisby\src\frisby\spec.js:290:26)
    at _fetch._fetch.then (C:\Data\Github\juice-shop\node_modules\frisby\src\frisby\spec.js:222:18)
    at process._tickCallback (internal/process/next_tick.js:109:7)' to be null.
```
### v2 (workaround)
```
  it('gets existing basket with contained products by id', function (done) {
    frisby.get(REST_URL + '/basket/1', {headers: authHeader})
      .expect('status', 200)
      .expect('header', 'content-type', /application\/json/)
      .then(function (json) {
        expect(json.data.id).toBe(1)
      })
      .done(done)
  })
```
fails as well, because what is supposed to be the JSON in the `json` param is instead:
```
FrisbyResponse {
  _response:
   Body {
     url: 'http://localhost:3000/rest/basket/1',
     status: 200,
     statusText: 'OK',
     headers: Headers { _headers: [Object] },
     ok: true,
     body:
      PassThrough {
        _readableState: [Object],
        readable: false,
        domain: null,
        _events: [Object],
        _eventsCount: 3,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: false,
        allowHalfOpen: true,
        _transformState: [Object] },
     bodyUsed: true,
     size: 0,
     timeout: 5000,
     _raw: [ <Buffer 7b 22 73 74 61 74 75 73 22 3a 22 73 75 63 63 65 73 73 22 2c 22 64 61 74 61 22 3a 7b 22 69 64 22 3a 31 2c 22 63 6f 75 70 6f 6e 22 3a 6e 75 6c 6c 2c 22 ... > ],
     _abort: false,
     _bytes: 1289 },
  _body:
   { status: 'success',
     data:
      { id: 1,
        coupon: null,
        createdAt: '2017-07-26T12:48:41.000Z',
        updatedAt: '2017-07-26T12:48:41.000Z',
        UserId: 1,
        products: [Object] } } }
```

I would prefer if `expect('json')` would work like `expectJSON()` did, but if that's not intended I need at least the workaround to work <or> a hint how to implement the test correctly in _v2-style_... :-)

--- a/spec/expects_jsonTypes_spec.js
+++ b/spec/expects_jsonTypes_spec.js
@@ -6,7 +6,7 @@ const mocks = require('./fixtures/http_mocks');
 
 const testHost = 'http://api.example.com';
 
-describe('Frisby expect jsonTypes', function() {
+describe('expect(\'jsonTypes\')', function() {
 
   it('should match exact JSON', function(doneFn) {
     mocks.use(['getUser1']);
@@ -155,6 +155,24 @@ describe('Frisby expect jsonTypes', function() {
         "foo": "bar"
       })
       .expect('jsonTypes', {
+        "name": Joi.string()
+      })
+      .done(doneFn);
+  });
+});
+
+describe('expect(\'jsonTypesStrict\')', function() {
+
+  it('should error on additional JSON keys not accounted for', function (doneFn) {
+    frisby.fromJSON({
+        "name": "john",
+        "foo": "bar"
+      })
+      .expect('jsonTypesStrict', {
+        "name": Joi.string(),
+        "foo": Joi.string()
+      })
+      .expectNot('jsonTypesStrict', {
         "name": Joi.string()
       })
       .done(doneFn);

--- a/spec/expects_json_spec.js
+++ b/spec/expects_json_spec.js
@@ -5,7 +5,7 @@ const mocks = require('./fixtures/http_mocks');
 
 const testHost = 'http://api.example.com';
 
-describe('Frisby expect json', function() {
+describe('expect(\'json\')', function() {
 
   it('should match exact JSON', function(doneFn) {
     mocks.use(['getUser1']);
@@ -23,79 +23,6 @@ describe('Frisby expect json', function() {
 
     frisby.fetch(testHost + '/users/1')
       .expectNot('json', {
-        id: 1,
-        id2: 2,
-        email: 'joe.schmoe@example.com'
-      })
-      .done(doneFn);
-  });
-
-  it('should error with missing key', function(doneFn) {
-    mocks.use(['getUser1']);
-
-    frisby.fetch(testHost + '/users/1')
-      .expectNot('json', {
-        email: 'joe.schmoe@example.com'
-      })
-      .done(doneFn);
-  });
-
-  it('should error with matching keys, but incorrect values', function(doneFn) {
-    mocks.use(['getUser1']);
-
-    frisby.fetch(testHost + '/users/1')
-      .expectNot('json', {
-        id: 1,
-        email: 'joe.schmoe@example.net'
-      })
-      .done(doneFn);
-  });
-
-  it('should match from data via fromJSON', function(doneFn) {
-    frisby.fromJSON({
-        foo: 'bar'
-      })
-      .expect('json', {
-        foo: 'bar'
-      })
-      .done(doneFn);
-  });
-
-  it('should match JSON in using provided path', function(doneFn) {
-    frisby.fromJSON({
-      one: {
-        two: {
-          three: 3
-        }
-      }
-    })
-    .expect('jsonContains', 'one.two', {
-      three: 3
-    })
-      .done(doneFn);
-  });
-
-});
-
-
-describe('Frisby expect jsonContains', function() {
-
-  it('should match exact JSON', function(doneFn) {
-    mocks.use(['getUser1']);
-
-    frisby.fetch(testHost + '/users/1')
-      .expect('jsonContains', {
-        id: 1,
-        email: 'joe.schmoe@example.com'
-      })
-      .done(doneFn);
-  });
-
-  it('should error with extra key', function(doneFn) {
-    mocks.use(['getUser1']);
-
-    frisby.fetch(testHost + '/users/1')
-      .expectNot('jsonContains', {
         id: 1,
         id2: 2,
         email: 'joe.schmoe@example.com'
@@ -107,7 +34,7 @@ describe('Frisby expect jsonContains', function() {
     mocks.use(['getUser1']);
 
     frisby.fetch(testHost + '/users/1')
-      .expect('jsonContains', {
+      .expect('json', {
         email: 'joe.schmoe@example.com'
       })
       .done(doneFn);
@@ -117,7 +44,7 @@ describe('Frisby expect jsonContains', function() {
     mocks.use(['getUser1']);
 
     frisby.fetch(testHost + '/users/1')
-      .expectNot('jsonContains', {
+      .expectNot('json', {
         id: 1,
         email: 'joe.schmoe@example.net'
       })
@@ -129,13 +56,13 @@ describe('Frisby expect jsonContains', function() {
         foo: 'bar',
         bar: 'baz'
       })
-      .expect('jsonContains', {
+      .expect('json', {
         foo: 'bar'
       })
       .done(doneFn);
   });
 
-  it('should error with extra nested keys', function(doneFn) {
+  it('should error with incorrect nested key value', function(doneFn) {
     frisby.fromJSON({
         one: {
           two: {
@@ -143,12 +70,83 @@ describe('Frisby expect jsonContains', function() {
           }
         }
       })
-      .expectNot('jsonContains', {
+      .expectNot('json', {
         one: {
           two: {
             three: 4
           }
         }
+      })
+      .done(doneFn);
+  });
+
+  it('should match JSON content using provided path and object', function(doneFn) {
+    frisby.fromJSON({
+      one: {
+        two: {
+          three: 3
+        }
+      }
+    })
+    .expect('json', 'one.two', {
+      three: 3
+    })
+      .done(doneFn);
+  });
+
+});
+
+describe('expect(\'jsonStrict\')', function() {
+  it('should match exact JSON', function(doneFn) {
+    mocks.use(['getUser1']);
+
+    frisby.fetch(testHost + '/users/1')
+      .expect('jsonStrict', {
+        id: 1,
+        email: 'joe.schmoe@example.com'
+      })
+      .done(doneFn);
+  });
+
+  it('should error with extra key', function(doneFn) {
+    mocks.use(['getUser1']);
+
+    frisby.fetch(testHost + '/users/1')
+      .expectNot('jsonStrict', {
+        id: 1,
+        id2: 2,
+        email: 'joe.schmoe@example.com'
+      })
+      .done(doneFn);
+  });
+
+  it('should error with missing key', function(doneFn) {
+    mocks.use(['getUser1']);
+
+    frisby.fetch(testHost + '/users/1')
+      .expectNot('jsonStrict', {
+        email: 'joe.schmoe@example.com'
+      })
+      .done(doneFn);
+  });
+
+  it('should error with matching keys, but incorrect values', function(doneFn) {
+    mocks.use(['getUser1']);
+
+    frisby.fetch(testHost + '/users/1')
+      .expectNot('jsonStrict', {
+        id: 1,
+        email: 'joe.schmoe@example.net'
+      })
+      .done(doneFn);
+  });
+
+  it('should match from data via fromJSON', function(doneFn) {
+    frisby.fromJSON({
+        foo: 'bar'
+      })
+      .expect('jsonStrict', {
+        foo: 'bar'
       })
       .done(doneFn);
   });

--- a/spec/frisby_spec.js
+++ b/spec/frisby_spec.js
@@ -75,9 +75,9 @@ describe('Frisby', function() {
     mocks.use(['getUser1', 'getUser2WithDelay']);
 
     frisby.get(testHost + '/users/1')
-      .expect('jsonContains', { id: 1 })
+      .expect('json', { id: 1 })
       .then(frisby.get(testHost + '/users/2')
-          .expect('jsonContains', { id: 2 })
+          .expect('json', { id: 2 })
       )
       .then(function (res) {
         expect(res.json.id).toBe(2);
@@ -89,10 +89,10 @@ describe('Frisby', function() {
     mocks.use(['getUser1', 'getUser2WithDelay']);
 
     frisby.get(testHost + '/users/1')
-      .expect('jsonContains', { id: 1 })
+      .expect('json', { id: 1 })
       .then(function () {
         return frisby.get(testHost + '/users/2')
-          .expect('jsonContains', { id: 2 });
+          .expect('json', { id: 2 });
       })
       .then(function (res) {
         expect(res.json.id).toBe(2);
@@ -104,19 +104,19 @@ describe('Frisby', function() {
     mocks.use(['getUser1', 'getUser2WithDelay']);
 
     frisby.get(testHost + '/users/1')
-      .expect('jsonContains', { id: 1 })
+      .expect('json', { id: 1 })
       .then(function () {
         mocks.use(['getUser1WithDelay']);
 
         return frisby.get(testHost + '/users/1')
-          .expect('jsonContains', { id: 1 });
+          .expect('json', { id: 1 });
       })
       .then(function (res) {
         expect(res.json.id).toBe(1);
       })
       .then(function () {
         return frisby.get(testHost + '/users/2')
-          .expect('jsonContains', { id: 2 });
+          .expect('json', { id: 2 });
       })
       .then(function (res) {
         expect(res.json.id).toBe(2);

--- a/src/frisby.js
+++ b/src/frisby.js
@@ -101,8 +101,20 @@ function removeExpectHandler(expectName, expectFn) {
 }
 
 module.exports = {
+  addExpectHandler,
   baseUrl,
-  del, fetch, fromJSON, get: get, globalSetup, Joi, patch, post, put, setup, timeout, use,
+  del,
+  fetch,
+  fromJSON,
+  get: get,
+  globalSetup,
+  Joi,
+  patch,
+  post,
+  put,
+  removeExpectHandler,
+  setup,
+  timeout,
+  use,
   version,
-  addExpectHandler, removeExpectHandler,
 };

--- a/src/frisby/expects.js
+++ b/src/frisby/expects.js
@@ -64,20 +64,20 @@ const expects = {
 
     incrementAssertionCount();
 
-    utils.withPath(path, response._body, function jsonAssertion(jsonChunk) {
-      assert.deepEqual(json, jsonChunk);
+    utils.withPath(path, response._body, function jsonContainsAssertion(jsonChunk) {
+      let failMsg = "Response [ " + JSON.stringify(jsonChunk) + " ] does not contain provided JSON [ " + JSON.stringify(json) + " ]";
+      assert.ok(_.some([jsonChunk], json), failMsg);
     });
   },
 
-  jsonContains(response, _path, _json) {
+  jsonStrict(response, _path, _json) {
     let json = _json ? _json : _path;
     let path = _json ? _path : false;
 
     incrementAssertionCount();
 
-    utils.withPath(path, response._body, function jsonContainsAssertion(jsonChunk) {
-      let failMsg = "Response [ " + JSON.stringify(jsonChunk) + " ] does not contain provided JSON [ " + JSON.stringify(json) + " ]";
-      assert.ok(_.some([jsonChunk], json), failMsg);
+    utils.withPath(path, response._body, function jsonAssertion(jsonChunk) {
+      assert.deepEqual(json, jsonChunk);
     });
   },
 
@@ -94,7 +94,22 @@ const expects = {
         throw result.error;
       }
     });
-  }
+  },
+
+  jsonTypesStrict(response, _path, _json) {
+    let json = _json ? _json : _path;
+    let path = _json ? _path : false;
+
+    incrementAssertionCount();
+
+    utils.withPath(path, response._body, function jsonTypesAssertion(jsonChunk) {
+      let result = Joi.validate(jsonChunk, json);
+
+      if (result.error) {
+        throw result.error;
+      }
+    });
+  },
 
 };
 

